src/app/auth/login.ts

```diff
@@ -113,6 +136,138 @@ export const loginWithPhoneCookie = async ({
+  //
+  // Scenario 2) Happy Path - User logs in with device jwt, no phone account exists, upgrade device to phone account
+  //             a. update kratos => update schema to phone_no_password_v0, remove device trait
+  //             b. mongo (user) => update user collection and remove device field, add phone
+  //             c. mongo (account) => update account to level 1
+  //
+  // Scenario 3) Unhappy path - User sells phone and forgets to sweep funds to phone account
+  //             Option 1 - too bad, we can't help them
+  //             Option 2 - create some kind of recovery code process?
+
+  // Scenario 1 - does phone account already exist?
+  const phoneAccount = await UsersRepository().findByPhone(phone)
+
+  if (phoneAccount instanceof CouldNotFindUserFromPhoneError) {
+    // nothing to do, we'll go to scenario 2
+  } else if (phoneAccount instanceof Error) {
+    return phoneAccount
+  } else {
+    // is there still txns left over on the device account?
+    const deviceWallets = await WalletsRepository().listByAccountId(account.id)
+    if (deviceWallets instanceof Error) return deviceWallets
+    const ledger = LedgerService()
+    let deviceAccountHasBalance = false
+    for (const wallet of deviceWallets) {
+      const balance = await ledger.getWalletBalance(wallet.id)
+      if (balance instanceof Error) return balance
+      if (balance > 0) {
+        deviceAccountHasBalance = true
+      }
+    }
+    if (deviceAccountHasBalance) {
+      // Scenario 1.5 - has txns on device account but phone account exists
+      return new PhoneAccountAlreadyExistsNeedToSweepFundsError()
+    } else {
+      // Scenario 1 - no txns on device account but phone account exists
+      // just log the user in with the phone account
+      const authService = AuthWithPhonePasswordlessService()
+      const kratosResult = await authService.loginToken({ phone })
+      if (kratosResult instanceof Error) return kratosResult
+      return kratosResult.sessionToken
+    }
+  }
+
+  // Scenario 2 - Happy Path
+  // a. create kratos account
+  // b. and c. migrate account/user collection in mongo via kratos/registration webhook
+  const kratosToken = await AuthWithDeviceAccountService().upgradeToPhoneSchema({
+    phone,
+    deviceId: account.kratosUserId,
+  })
+  if (kratosToken instanceof Error) return kratosToken
+  return kratosToken
+}
+
+export const loginWithDevice = async ({
+  jwt,
+  ip,
+}: {
+  jwt: string
+  ip: IpAddress
+}): Promise<SessionToken | ApplicationError> => {
+  {
+    const limitOk = await checkFailedLoginAttemptPerIpLimits(ip)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  const verifiedJwt = await verifyJwt(jwt)
+  if (verifiedJwt instanceof Error) {
+    await rewardFailedLoginAttemptPerIpLimits(ip)
+    return verifiedJwt
+  }
+  const deviceId = verifiedJwt.sub as UserId
+  if (!deviceId) return new JwtSubjectUndefinedError("deviceId sub is undefined")
+
+  // 1. Does account exist in mongo?
```

Comment:
why do we need a comment on something that is evident?

Reply:
forget to delete it. it helps prompt github co-pilot to write the code for me lol